cirq/experiments/cross_entropy_benchmarking.py

```diff
@@ -0,0 +1,347 @@
+        number of cycles and the second item being the corresponding average
+        XEB fidelity.
+        """
+        return [(num, prob) for num, prob in zip(self._num_cycles,
+                                                 self._xeb_fidelities)]
+
+    def plot(self, **plot_kwargs: Any) -> None:
+        """Plots the average XEB fidelity vs the number of cycles.
+
+        Args:
+            **plot_kwargs: Arguments to be passed to matplotlib.pyplot.plot.
+        """
+        fig = plt.figure()
+        ax = plt.gca()
+        ax.set_ylim([0, 1.1])
+
+        plt.plot(self._num_cycles, self._xeb_fidelities, 'ro-',
+                 figure=fig, **plot_kwargs)
+        plt.xlabel('Number of Cycles', figure=fig)
+        plt.ylabel('XEB Fidelity', figure=fig)
+        fig.show()
+
+
+def cross_entropy_benchmarking(
+        sampler: sim.Sampler,
+        qubits: Sequence[devices.GridQubit],
+        two_qubit_gate: ops.gate_features.TwoQubitGate,
+        *,
+        num_circuits: int = 20,
+        repetitions: int = 1000,
+        num_cycle_range: Sequence[int] = range(2, 103, 10),
+        interaction_sequence: List[Set[Tuple[devices.GridQubit,
+                                             devices.GridQubit]]] = None,
+        use_tetrahedral_group: bool = False) -> CrossEntropyResult:
+    r"""Cross-entropy benchmarking (XEB) of multiple qubits.
+
+    A total of M random circuits are generated, each of which is made of N
+    cycles where N = max(num_cycle_range). Every cycle contains randomly
+    generated single-qubit gates applied to each qubit, followed by two-qubit
+    gate(s) applied to one or more pairs of qubits. For each given number of
+    cycles n in num_cycle_range, the experiment performs the following:
+
+    1) Measure the bit-string probabilities P^(meas, m)_|...00>,
+    P^(meas, m)_|...01>, P^(meas, m)_|...10>, P^(meas, m)_|...11>... at the
+    end of circuit_mn for all m, where circuit_mn is the first n cycles of
+    random circuit m (m <= M).
+
+    2) Theoretically compute the expected bit-string probabilities
+    P^(exp, m)_|...00>,  P^(exp, m)_|...01>, P^(exp, m)_|...10>,
+    P^(exp, m)_|...11>... at the end of circuit_mn for all m.
+
+    3) Collect all measured probabilities into a single vector P_meas and
+    normalize P_meas such it sums to 1.
+
+    4) Collect all expected probabilities into a single vector P_exp. Set any
+    probabilities <1e-22 to 1e-22 to avoid singularities in log P_exp. Then
+    normalize P_exp such it sums to 1.
+
+    5) Calculate the "XEB fidelity" at this given circuit depth using the
+    following expressions:
+
+    s_incoherent = -np.sum(P_uniform * np.log(P_exp))
+    s_expected = -np.sum(P_exp * np.log(P_exp))
+    s_meas = -np.sum(P_meas * np.log(P_exp))
+    fidelity = (s_incoherent - s_meas) / (s_incoherent - s_expected)
+
+    where P_uniform = 1/(M * 2 ** num_qubits) is a constant probability
```

Comment:
Why is M in this equation?

Reply:
M is here because the vectors `P_exp`, `P_meas` and `P_uniform` include bit-string probabilities from all M random circuits, so to normalize each of them there is an additional division by M. I clarified this by including sentences in the doc-string stating this additional normalization.